docs/notebooks/Hidden_Linear_Function_Problem.ipynb

```diff
@@ -0,0 +1,527 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Hidden Linear Function Problem\n",
+    "\n",
+    "In this notebook we consider a problem from the paper \"[Quantum advantage with shallow circuits](https://arxiv.org/abs/1704.00690)\" and build a quantum circuit, which solves it, in Cirq. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "### The problem\n",
+    "\n",
+    "In this problem we consider a [quadratic form](https://en.wikipedia.org/wiki/Quadratic_form) of a [binary vector](https://en.wikipedia.org/wiki/Bit_array) and with binary coefficients (but additions and multiplications are evaluated modulo 4). Then we restrict this quadratic form, i.e. we allow to use only certain binary vectors as input. It turns out that under this restriction this quadratic form is equivalent to a linear function, i.e. it just evaluates dot product of input vector and certain scalar vector. Task is to find this scalar vector.\n",
+    "\n",
+    "In other words, we have linear function, which is \"hidden\" inside a quadratic form.\n",
+    "\n",
+    "### Formal statement of the problem\n",
+    "\n",
+    "Consider $A \\in \\mathbb{F}_2^{n \\times n}$ - upper-triangular binary matrix, $b \\in \\mathbb{F}_2^n$ - binary vector.\n",
+    "\n",
+    "Define a function $q : \\mathbb{F}_2^n \\to \\mathbb{Z}_4 $:\n",
+    "\n",
+    "$$q(x) = (2 x^T A x + b^T x) ~\\text{mod}~ 4, $$ \n",
+    "\n",
+    "Also define\n",
+    "\n",
+    "$$\\mathcal{L}_q = \\Big\\{x \\in  \\mathbb{F}_2^n : q(x \\oplus y) = (q(x) + q(y)) ~\\text{mod}~ 4 ~~ \\forall y \\in \\mathbb{F}_2^n \\Big\\}.$$\n",
+    "\n",
+    "Turns out (see Lemma 1 on page 6 in [1]) that restriction of $q$ on $\\mathcal{L}_q$ is a linear function, i.e. there exists such $z \\in \\mathbb{F}_2^n$, that\n",
+    "\n",
+    "$$q(x) = 2 z^T x ~~\\forall x \\in \\mathcal{L}_q.$$\n",
+    "\n",
+    "Our task is, given $A$ and $b$, to find $z$. There may be multiple answers - we need to find any such answer.\n",
+    "\n",
+    "### Why is this problem interesting?\n",
+    "\n",
+    "#### 1. It's a problem without an oracle\n",
+    "\n",
+    "This problem is similar to a problem solved by [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm). It also finds coefficients of unknown linear function. But in the Bernstein-Vazirani algorithm this function is represented by an oracle. In this problem, the linear function is \"hidden\" in inputs $A$ and $b$.\n",
```

Comment:
I think this paragraph needs some rewording since it is a bit awkward.  Here's my suggestion:

The [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm) also finds coefficients of an unknown linear function by representing by it as an oracle. By hiding the linear function in the inputs $A$ and $b$.\n", we avoid the use of an oracle.

You might also want to add why eliminating the oracle is useful.

Reply:
> You might also want to add why eliminating the oracle is useful.

Yes. We really want to emphasize why this problem is important: It shows an unconditional separation between quantum and classical complexity classes. A result we usually don't get to have. In this case, a search problem in QNC^0 that is not in NC^0 (Link to https://en.wikipedia.org/wiki/NC_(complexity))

Reply:
Rephrased.
Added the fact about complexity classes to next section about "Quantum circuits have advantage over classical when solving this problem"